utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,602 @@
+makeVertexInputBindingDescription (binding, stride) = zero
+  { binding   = fromIntegral binding
+  , stride    = fromIntegral stride
+  , inputRate = VERTEX_INPUT_RATE_VERTEX
+  }
+
+makeVertexAttribute :: Input -> Vector VertexAttribute
+makeVertexAttribute Input {..} = V.map
+  (\i -> VertexAttribute { binding  = 0
+                         , location = fromIntegral . (+ location) $ i
+                         , size
+                         , format
+                         }
+  )
+  [0 .. count - 1]
+ where
+  count          = maybe 1 V.sum array :: Int
+  (size, format) = convertVertexAttributeType type'
+
+makeVertexInputAttributeDescriptions
+  :: Vector VertexAttribute -> Vector VertexInputAttributeDescription
+makeVertexInputAttributeDescriptions =
+  V.fromList
+    . join
+    . map process
+    . groupBy ((==) `on` (binding :: VertexAttribute -> Word32))
+    . V.toList
+ where
+  process :: [VertexAttribute] -> [VertexInputAttributeDescription]
+  process = snd . foldl
+    (\(nextOffset, acc) cur -> liftA2
+      (,)
+      fst
+      ((acc ++) . pure . snd)
+      (makeVertexInputAttributeDescription nextOffset cur)
+    )
+    (0, [])
+
+makeVertexInputAttributeDescription
+  :: Word32 -> VertexAttribute -> (Word32, VertexInputAttributeDescription)
+makeVertexInputAttributeDescription offset VertexAttribute {..} =
+  (offset + size, zero { binding, location, offset, format })
+
+reflect
+  :: MonadIO m
+  => ShaderStage
+  -> "code" ::: Text
+  -> m (B.ByteString, BL.ByteString)
+reflect shaderStage code =
+  liftIO . withSystemTempDirectory "th-spirv" $ \dir -> do
+    let stage  = T.unpack . shaderStageToText $ shaderStage
+    let shader = dir </> "glsl." <> stage
+    let spv    = dir </> stage <> ".spv"
+    T.writeFile shader code
+    (_exitCode, _spv, _err) <-
+      readProcess
+      . proc "glslangValidator"
+      $ ["-S", stage, "-V", shader, "-o", spv]
+    spirv                            <- B.readFile spv
+    (_exitCode, reflectionRaw, _err) <-
+      readProcess
+      . proc "spirv-cross"
+      $ [spv, "--vulkan-semantics", "--reflect"]
+    pure (spirv, reflectionRaw)
+
+reflect' :: MonadIO m => "spirv" ::: B.ByteString -> m BL.ByteString
+reflect' spirv = liftIO . withSystemTempDirectory "th-spirv" $ \dir -> do
+  let spv = dir </> "glsl.spv"
+  B.writeFile spv spirv
+  (_exitCode, reflectionRaw, _err) <-
+    readProcess . proc "spirv-cross" $ [spv, "--vulkan-semantics", "--reflect"]
+  pure reflectionRaw
+
+reflection
+  :: MonadIO m => ShaderStage -> "code" ::: Text -> m (Shader, Reflection)
+reflection stage code = do
+  (spirv, reflectionRaw) <- reflect stage code
+  case decode reflectionRaw of
+    Just reflection -> pure (Shader { stage, code = spirv }, reflection)
+    Nothing         -> error "fail to reflect"
+
+reflection' :: MonadIO m => "spirv" ::: B.ByteString -> m (Shader, Reflection)
+reflection' spirv = do
+  reflectionRaw <- reflect' spirv
+  case decode reflectionRaw of
```

Comment:
`eitherDecodeStrict'` and pass the error message, a user would want to see what happened.